src/v1/Models/Config/Role.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+/**
+ * FusionSuite - Backend
+ * Copyright (C) 2022 FusionSuite
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace App\v1\Models\Config;
+
+use Illuminate\Database\Eloquent\Model as Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
+use Illuminate\Database\Capsule\Manager as DB;
+use Illuminate\Database\Eloquent\Builder;
+
+class Role extends Model
+{
+  use SoftDeletes;
+
+  protected $appends = [
+    'users',
+    'permissiondatas',
+    'permissionstructures'
+  ];
+
+  protected $visible = [
+    'id',
+    'name',
+    'permissionstructure',
+    'permissiondata',
+    'users',
+    'permissiondatas',
+    'permissionstructures',
+    'created_at',
+    'updated_at',
+    'deleted_at',
+    'created_by',
+    'updated_by',
+    'deleted_by'
+  ];
+
+  public static function boot()
+  {
+    parent::boot();
+    static::creating(function ($model)
+    {
+      $model->created_by = $GLOBALS['user_id'];
+    });
+
+    static::updating(function ($model)
+    {
+      $model->updated_by = $GLOBALS['user_id'];
+    });
+
+    static::deleting(function ($model)
+    {
+      // Disable timestamp to prevent updating updated_at field when soft delete
+      $model->timestamps = false;
+      // Update deleted_by to have user id
+      DB::table('roles')->where('id', $model->id)
+        ->update(['deleted_by' => $GLOBALS['user_id']]);
+    });
+
+    static::restoring(function ($model)
+    {
+      $model->deleted_by = null;
+    });
+  }
+
+  public function getUsersAttribute()
+  {
+    $userList = [];
+    $users = $this->users()->get();
+    foreach ($users as $user)
+    {
+      $userList[] = \App\v1\Models\Common::getUserAttributes($user->id);
+    }
+    return $userList;
+  }
+
+  public function getCreatedByAttribute($value)
+  {
+    return \App\v1\Models\Common::getUserAttributes($value);
+  }
+
+  public function getUpdatedByAttribute($value)
+  {
+    return \App\v1\Models\Common::getUserAttributes($value);
+  }
+
+  public function getDeletedByAttribute($value)
+  {
+    return \App\v1\Models\Common::getUserAttributes($value);
+  }
```

Comment:
I don't understand these methods. Did you want to do something like this?

```suggestion
  public function getCreatedByUserAttribute()
  {
    return \App\v1\Models\Common::getUserAttributes($this->created_by);
  }

  public function getUpdatedByUserAttribute()
  {
    return \App\v1\Models\Common::getUserAttributes($this->updated_by);
  }

  public function getDeletedByUserAttribute()
  {
    return \App\v1\Models\Common::getUserAttributes($this->deleted_by);
  }
```

Reply:
It's the system of the eloquent model, the function name is a norme base on field name